diff --git a/gcc/Makefile.in b/gcc/Makefile.in
index 2c3f02e8953..9ca6e6bbd44 100644
--- a/gcc/Makefile.in
+++ b/gcc/Makefile.in
@@ -1242,16 +1242,6 @@ OBJS = \
 	godump.o \
 	graph.o \
 	graphds.o \
-	graphite.o \
-	graphite-blocking.o \
-	graphite-clast-to-gimple.o \
-	graphite-isl-ast-to-gimple.o \
-	graphite-dependences.o \
-	graphite-interchange.o \
-	graphite-optimize-isl.o \
-	graphite-poly.o \
-	graphite-scop-detection.o \
-	graphite-sese-to-poly.o \
 	gtype-desc.o \
 	haifa-sched.o \
 	hw-doloop.o \
diff --git a/gcc/passes.def b/gcc/passes.def
index 2305d670663..968454ce887 100644
--- a/gcc/passes.def
+++ b/gcc/passes.def
@@ -232,13 +232,6 @@ along with GCC; see the file COPYING3.  If not see
 	  NEXT_PASS (pass_check_data_deps);
 	  NEXT_PASS (pass_loop_distribution);
 	  NEXT_PASS (pass_copy_prop);
-	  NEXT_PASS (pass_graphite);
-	  PUSH_INSERT_PASSES_WITHIN (pass_graphite)
-	      NEXT_PASS (pass_graphite_transforms);
-	      NEXT_PASS (pass_lim);
-	      NEXT_PASS (pass_copy_prop);
-	      NEXT_PASS (pass_dce);
-	  POP_INSERT_PASSES ()
 	  NEXT_PASS (pass_iv_canon);
 	  NEXT_PASS (pass_parallelize_loops);
 	  NEXT_PASS (pass_if_conversion);
