diff --git a/gcc/graphite-optimize-isl.c b/gcc/graphite-optimize-isl.c
index c1d04af..0ef5157 100644
--- a/gcc/graphite-optimize-isl.c
+++ b/gcc/graphite-optimize-isl.c
@@ -354,7 +356,7 @@ getScheduleForBandList (isl_band_list *BandList)
 	{
 	  for (i = ScheduleDimensions - 1 ;  i >= 0 ; i--)
 	    {
-	      if (isl_band_member_is_zero_distance (Band, i))
+	      if (isl_band_member_is_coincident (Band, i))
 		{
 		  isl_map *TileMap;
 		  isl_union_map *TileUMap;
